Mac compatibility: Avoid PCRE grep in recon-surf, stat formatting and mapfile

dkuegler opened the pull request:
Avoid using grep -P, instead use grep -E and some bash instructions.

This PR addresses #668 and was introduced by #652.

Needs testing to see if grep 2.6 / mac versions of grep on mac support the extended grep syntax.

This PR also addresses 2 more Mac compatibility issues:
- The use of `mapfile` in run_fastsurfer and brun_fastsurfer (the latter will just not be compatible for some specific flags/arguments)
- the use of `stat` which uses `-c` on Linux (GNU coreutils version) and `-f` for similar things on macOS (BSD version)

dkuegler committed:
Avoid PCRE grep in recon-surf

Avoid using grep -P, instead use grep -E and some bash instructions.

m-reuter commented:
This works on my Mac. I am getting a print out
```A min-conformedconformed image ... ``` which looks weird. 

Also I wonder if we should use this PR to fix the other Mac issues:
- the BSD tool ```stat``` in Mac does not have the -c flag (I think it uses -f flag for format)
- the missing ```mapfile```

dkuegler commented:
> This works on my Mac. I am getting a print out `A min-conformedconformed image ... ` which looks weird.
> 
> Also I wonder if we should use this PR to fix the other Mac issues:
> 
>     * the BSD tool `stat` in Mac does not have the -c flag (I think it uses -f flag for format)
> 
>     * the missing `mapfile`

Let's use this issue for that. The grep issue I was fixing here is a mac-only issue.

I will fix the `mapfile` thing very quickly (this is just an information things, so no damage, just going to skip that for old bash versions).
I will see about `stat` as well

dkuegler committed:
Fix all uses of mapfile, i.e. gate them behind an error message or deactivate the part if it is optional.

dkuegler commented:
Still working on stat
`mapfile` changes are untested on macos

m-reuter commented:
I tested mapfile changes and it works (also tested --threads 4 for parallel). 

One other thing, I get the run time only in full hours : ```#@#%# recon-surf-run-time-hours  0.000```

Probably due to first output line: ```time command failing, not using time...```
while a time command is available on Mac, probably uses a different format. 
Update: I just saw that on Linux it also only reports the full hours. So we should probably report also minutes in both cases and we should check if the time command failing printout should be changed to a message that looks less like an error or warning, maybe "unexpected time command (maybe on MAC), using ??? instead"

dkuegler committed:
Move the check for subjects_dir (that is repeated across different files) into a function in functions.
This check uses `stat`, which works in different ways for macOS and linux.
Make the function (stat) work for both linux (GNU coreutils) and macOS (unclear version).

dkuegler committed:
Modify the time check message (not "fail")
use bc -l flag to use fractions as well

dkuegler commented on recon_surf/functions.sh:
```diff
@@ -1,22 +1,48 @@
+    if [[ "$(stat --version | head -n 1)" =~ GNU[[:space:]]coreutils ]] ; then # linux
+      user_group=$(stat -c "%u:%g" "$1")
+      world_access=$(stat -c "%a" "$1" | tail -c 2)
+    elif [[ "$(stat -- 2>&1 | tail -n 1)" =~ ^usage:[[:space:]]stat.*-f[[:space:]]format ]] ; then # macOS
```
@m-reuter this is the line where `2>&1` does not capture stderr it seems

m-reuter replied:
```
> stat --version
stat: illegal option -- -
usage: stat [-FLnq] [-f format | -l | -r | -s | -x] [-t timefmt] [file ...]
```

m-reuter replied:
also the whole block gives the same error message and stops, does not go into elif block.

m-reuter replied:
I think this way would work:
```
if stat --version &>/dev/null; then
  echo "This is the GNU version of stat."
else
  echo "This is the BSD version of stat."
fi
```

m-reuter commented:
I still get ```./FastSurfer/run_fastsurfer.sh: line 828: mapfile: command not found```. Running a test now.

dkuegler commented:
> I still get `./FastSurfer/run_fastsurfer.sh: line 828: mapfile: command not found`. Running a test now.

What do you get, if you run `bash --version` ?

m-reuter commented:
inside the scripts it uses the system bash, which is this:
```
> /bin/bash --version
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin24)
Copyright (C) 2007 Free Software Foundation, Inc.
```
in my terminal I have a newer one (5.something) via home-brew, but system bash is still old, and that is the one that gets used via the bash shebang.

dkuegler commented:
So the difference is that when we call `bash --version`, it is different to `./reconsurf.sh` which uses the other bash .... This is so confusing.

m-reuter commented:
it can be different depending on what other bash is installed on the system and in the path. On normal systems, it should be the same.

dkuegler commented:
I have changed the usage of `bash --version` to `${BASH_VERSINFO[0]}` which is available in bash 3.2 according to the docs of bash 3.2.

dkuegler committed:
remove version function in recon-surf and do not use awk to process the bash version number.
